scapy/layers/netflow.py

```diff
@@ -341,45 +383,87 @@ def dispatch_hook(cls, _pkt=None, *args, **kargs):
+    plist: the list of mixed NetflowV9 packets."""
+    # We need the whole packet to be dissected to access field def in NetflowFlowsetV9 or NetflowOptionsFlowsetV9
+    packet_list = plist.filter(lambda x: x.haslayer(NetflowFlowsetV9) or x.haslayer(NetflowOptionsFlowsetV9))
+    # Iterate through initial list
+    for pkt in plist:
         root = pkt.firstlayer()
-        current = root
+        if not NetflowDataflowsetV9 in root: # Restrict to NetflowDataflowsetV9
+            continue
         # Get all linked NetflowFlowsetV9
-        while current.payload.haslayer(NetflowFlowsetV9):
-            current = current.payload[NetflowFlowsetV9]
-            for ntv9 in current.templates:
-                current_ftl = root.getlayer(NetflowDataflowsetV9, templateID=ntv9.templateID)
+        for p in packet_list:
+            if NetflowFlowsetV9 in p: ## STEP 1 - NetflowFlowsetV9
+                current = p[NetflowFlowsetV9]
+                for ntv9 in current.templates:
+                    current_ftl = root.getlayer(NetflowDataflowsetV9, templateID=ntv9.templateID)
+                    if current_ftl:
+                        # Matched
+                        if len(current_ftl.records) < 0 or (not hasattr(current_ftl.records[0], "fieldValue")):
+                            # not necessary
+                            continue
+                        # All data is stored in one record, awaiting to be splitted
+                        data = current_ftl.records.pop(0).fieldValue
+                        res = []
+                        # Now, according to the NetflowFlowsetV9 data, re-dissect NetflowDataflowsetV9
+                        lengths_list = []
+                        for template in ntv9.template_fields:
+                            lengths_list.append((template.fieldLength, template.fieldType))
+                        if lengths_list:
+                            tot_len = sum(x for x,y in lengths_list)
+                            cls = _GenNetflowRecordV9(NetflowRecordV9, lengths_list)
+                            while len(data) >= tot_len:
+                                res.append(cls(data[:tot_len]))
+                                data = data[tot_len:]
+                        # Inject dissected data
+                        current_ftl.records = res
+                        current_ftl.do_dissect_payload(data)
+                        break
+            if NetflowOptionsFlowsetV9 in p: ## STEP 2 - NetflowOptionsFlowsetV9
+                current = p[NetflowOptionsFlowsetV9]
+                current_ftl = root.getlayer(NetflowDataflowsetV9, templateID=current.templateID)
                 if current_ftl:
                     # Matched
-                    if len(current_ftl.records) > 1:
-                        # post_dissection is not necessary
-                        return
+                    if len(current_ftl.records) < 0 or (not hasattr(current_ftl.records[0], "fieldValue")):
```

Comment:
Maybe you could `try...except` here rather than using `hashattr()`

Reply:
Good idea